### ISSUE TITLE
Trainer detect_anomaly may considerably slowdown performance

### ISSUE THREAD

**Issue by amit-miller** [PyTorchLightning/pytorch-lightning#12344]
Enabling the Trainer option `detect_anomaly` may cause very significant slow training.
In my case, I've experienced 3X slow down on various GPU platform [and widely fluctuating GPU utilizations].

I suggest adding an explicit warning to the documentation s.t. this option is used with care.

**Comment by rohitgr7** [comment]
this is used only for debugging. The link for this behavior is already here: https://pytorch-lightning.readthedocs.io/en/latest/common/debugging.html#detect-anomaly

**Comment by ananthsub** [comment]
The Lightning docs point to the PyTorch docs which explicitly mention this too: https://pytorch.org/docs/stable/autograd.html#anomaly-detection

> This mode should be enabled only for debugging as the different tests will slow down your program execution.

**Comment by amit-miller** [comment]
Deferring to a pytorch warning is not enough in this case. I strongly suggest you add an explicit text there.

**Comment by senarvi** [comment]
Usually you still want to terminate training, even when you're not debugging, if the loss explodes, right? `terminate_on_nan` checks whether the loss is infinite or NaN, but is deprecated and will be removed in the next release. [The documentation](https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html) suggests to use `detect_anomaly` instead. It doesn't say that one should use `detect_anomaly` just for debugging, so this is a very valid point. Also, there's no suggestion what to do if you want the `terminate_on_nan` behaviour always, not just when debugging.
